File: src/services/speedtest.ts
// ============================================================================
// services/speedtest.ts — Renderer client for the Rust-native speed test.
//
// The engine, the concurrency pool and the timeout budget all live in
// `src-tauri/src/core/speedtest.rs`. This module is only the thin IPC face:
//
//   startSpeedTest()  → fire and forget; returns the run id at once
//   cancelSpeedTest() → supersede the run owning a group
//   onDelayBatch()    → incremental results, batched by the engine
//   onDelayDone()     → exactly one terminal event per run, always
//
// No results travel back through the `invoke` return value. A few hundred
// nodes would otherwise have to be marshalled into a single payload, which is
// precisely the "wait for everything, then render" pattern this design exists
// to avoid.
//
// Like `services/tun.ts`, the payload types are declared here rather than
// pulled from the generated `@/bindings`: the speed-test commands are plain
// `#[tauri::command]`s (see the note in `commands/speedtest.rs` for why), so
// there is no generated type to import. Field names are camelCase because the
// Rust DTOs carry `#[serde(rename_all = "camelCase")]`.
// ============================================================================

import { safeInvoke, safeListen, type UnlistenFn } from '@/utils/tauri-bridge'

// ---------------------------------------------------------------------------
// Event names — must match the constants in `src-tauri/src/events.rs`.
// ---------------------------------------------------------------------------

export const DELAY_BATCH_EVENT = 'proxy://delay-batch'
export const DELAY_DONE_EVENT = 'proxy://delay-done'

// ---------------------------------------------------------------------------
// Defaults — must match the `const`s in `src-tauri/src/core/speedtest.rs`.
// Duplicated deliberately: the frontend needs them to seed its own state before
// the first event lands.
// ---------------------------------------------------------------------------

export const DEFAULT_TEST_URL = 'http://www.gstatic.com/generate_204'
export const DEFAULT_TIMEOUT_MS = 5_000
export const DEFAULT_CONCURRENCY = 64

// ---------------------------------------------------------------------------
// Wire types (mirror `core/speedtest.rs`)
// ---------------------------------------------------------------------------

/** Terminal outcome of one probe. */
export type ProbeStatus = 'ok' | 'timeout' | 'unreachable' | 'error'

export interface NodeProbe {
  name: string
  status: ProbeStatus
  /** RTT in ms; non-null iff `status === 'ok'`. */
  delayMs: number | null
  /** Why a non-`ok` probe ended that way. */
  message: string | null
}

/** Incremental progress push — one event per batch of results. */
export interface DelayBatch {
  runId: number
  group: string
  results: NodeProbe[]
  /** Nodes probed so far, including this batch. */
  done: number
  total: number
}

/** Terminal push — exactly one per run, including a cancelled run. */
export interface DelayDone {
  runId: number
  group: string
  total: number
  /** How many were actually probed; `< total` iff `cancelled`. */
  done: number
  ok: number
  failed: number
  cancelled: boolean
}

export interface StartSpeedTestOptions {
  url?: string
  timeoutMs?: number
  concurrency?: number
}

// ---------------------------------------------------------------------------
// Commands
// ---------------------------------------------------------------------------

/**
 * Kick off a run against `nodes` and return its id.
 *
 * Resolves as soon as the engine has accepted the run — not when the probes
 * finish. Results arrive on {@link DELAY_BATCH_EVENT}.
 *
 * Rejects on a genuinely bad request (e.g. an empty node list, which would
 * otherwise be a run that never emits a terminal event).
 */
export async function startSpeedTest(
  group: string,
  nodes: string[],
  opts: StartSpeedTestOptions = {},
): Promise<number> {
  return await safeInvoke<number>('speed_test_group', {
    group,
    nodes,
    url: opts.url ?? null,
    timeoutMs: opts.timeoutMs ?? null,
    concurrency: opts.concurrency ?? null,
  })
}

/**
 * Cancel the run owning `group`. Idempotent — cancelling an idle group is a
 * successful no-op, because the alternative (throwing) would make every
 * teardown path guard for a condition that is not an error.
 */
export async function cancelSpeedTest(group: string): Promise<void> {
  await safeInvoke<null>('cancel_speed_test', { group })
}

// ---------------------------------------------------------------------------
// Stream subscriptions
// ---------------------------------------------------------------------------

/** Subscribe to incremental results. Returns the unsubscribe handle. */
export function onDelayBatch(
  handler: (batch: DelayBatch) => void,
): Promise<UnlistenFn> {
  return safeListen<DelayBatch>(DELAY_BATCH_EVENT, (e) => handler(e.payload))
}

/** Subscribe to the per-run terminal event. */
export function onDelayDone(
  handler: (done: DelayDone) => void,
): Promise<UnlistenFn> {
  return safeListen<DelayDone>(DELAY_DONE_EVENT, (e) => handler(e.payload))
}

File: src-tauri/src/commands/speedtest.rs
// ============================================================================
// commands/speedtest.rs — Tauri surface for the Rust-native speed-test engine.
//
// Thin on purpose: the whole engine lives in `core::speedtest`, and this module
// only translates IPC into a `spawn_run` call. Two commands, both non-blocking:
//
//   speed_test_group   → returns the run id immediately, results stream back
//                        over `proxy://delay-batch` / `proxy://delay-done`
//   cancel_speed_test  → supersedes the run owning a group
//
// Why the command returns immediately instead of awaiting the pool: a
// few-hundred-node run takes tens of seconds, and the caller must be able to
// paint the first results long before the last probe lands. Awaiting here would
// also make the run's lifetime hostage to a single IPC call that a route change
// could abandon.
//
// NOTE ON BINDINGS: these commands are intentionally not registered with
// `tauri-specta` (no `#[specta]`, absent from `collect_commands!` in `lib.rs`).
// They follow the plain-command convention used by `tun` / `profile` / `proxy`,
// with the payload types declared on the renderer side in
// `src/services/speedtest.ts`. The reason is that the interesting types here are
// the *event* payloads, and specta types the command signature rather than the
// events — so registering would buy a typed `speedTestGroup` while the
// `DelayBatch` / `DelayDone` shapes stayed hand-written anyway. Not worth
// forcing a `src/bindings.ts` regeneration (which only happens when the GUI
// starts) for half the benefit.
// ============================================================================

use std::sync::Arc;

use tauri::{AppHandle, Manager};

use crate::core::speedtest::SpeedTestRegistry;
use crate::error::AppError;

type CmdResult<T> = Result<T, AppError>;

/// Resolve the managed registry.
///
/// `Arc<SpeedTestRegistry>` is the managed type so the spawned pool can hold an
/// owned handle instead of borrowing `State<'_, _>` across an `'static` future.
fn registry(app: &AppHandle) -> CmdResult<Arc<SpeedTestRegistry>> {
    app.try_state::<Arc<SpeedTestRegistry>>()
        .map(|s| s.inner().clone())
        .ok_or_else(|| AppError::Other("SpeedTestRegistry not registered".into()))
}

/// Probe every node in `nodes` concurrently and stream the results.
///
/// * `group`       — the owning group; used for cancellation and so the UI can
///                   route a batch to the right card set.
/// * `nodes`       — node names to probe. The caller filters out the
///                   un-probeable types (`Direct`, `Reject`, …) because only it
///                   knows the proxy tree.
/// * `url`         — probe target; defaults to the 204 endpoint.
/// * `timeout_ms`  — Mihomo's per-probe budget; defaults to 5000.
/// * `concurrency` — in-flight cap; defaults to 64, clamped by the engine.
///
/// Returns the run id. Every result carries it, so a client that superseded a
/// run can drop the stragglers — and so can we.
#[tauri::command]
#[specta::specta]
pub async fn speed_test_group(
    app: AppHandle,
    group: String,
    nodes: Vec<String>,
    url: Option<String>,
    timeout_ms: Option<u32>,
    concurrency: Option<u32>,
) -> CmdResult<u32> {
    let registry = registry(&app)?;

    // Empty input is a legitimate no-op, but returning a run id for a run that
    // will never emit anything would leave the caller waiting for a `done`
    // event that cannot come. Fail loudly instead — the UI never sends an
    // empty list, so this is a programming error, not a user-facing path.
    if nodes.is_empty() {
        return Err(AppError::Other(
            "speed_test_group called with no nodes".into(),
        ));
    }

    Ok(crate::core::speedtest::spawn_run(
        app,
        registry,
        group,
        nodes,
        url,
        timeout_ms,
        concurrency,
    ))
}

/// Cancel the run currently owning `group`.
///
/// Idempotent: cancelling an idle group is a successful no-op. The run stops at
/// its next probe completion and still emits a terminal
/// `proxy://delay-done` with `cancelled: true`, so the UI always gets to clear
/// its spinners.
#[tauri::command]
#[specta::specta]
pub fn cancel_speed_test(app: AppHandle, group: String) -> CmdResult<()> {
    registry(&app)?.cancel(&group);
    Ok(())
}
